File: python/verifier.py
import json
from .domain import Hypothesis, Evidence, HypothesisStatus
from .llm import ILLMProvider

class IndependentVerifier:
    def __init__(self, llm: ILLMProvider):
        self.llm = llm

    def verify_evidence(self, hypothesis: Hypothesis, evidence: Evidence) -> HypothesisStatus:
        prompt = f"""
You are an independent, objective software verification judge.
Evaluate whether the following repository evidence objectively supports or refutes the hypothesis.

Hypothesis:
- Statement: {hypothesis.statement}
- Location: {hypothesis.likely_source_location}
- Reasoning: {hypothesis.reasoning}

Evidence Gathered from Repository:
- Command: {evidence.source}
- Output:
{evidence.content[:1500]}

Respond ONLY with a JSON object:
{{
  "status": "SUPPORTED" | "REJECTED" | "INCONCLUSIVE"
}}
"""
        response = self.llm.generate(
            system_prompt="You are an independent code verification judge. Respond strictly in JSON.",
            user_prompt=prompt
        )

        cleaned = response.strip()
        if cleaned.startswith("```json"):
            cleaned = cleaned[7:]
        if cleaned.startswith("```"):
            cleaned = cleaned[3:]
        if cleaned.endswith("```"):
            cleaned = cleaned[:-3]
        cleaned = cleaned.strip()

        try:
            res = json.loads(cleaned)
            status_str = res.get("status", "INCONCLUSIVE").upper()
            if status_str in HypothesisStatus.__members__:
                return HypothesisStatus(status_str)
        except Exception:
            pass

        return HypothesisStatus.SUPPORTED if evidence.supports else HypothesisStatus.INCONCLUSIVE

File: src/verifier.ts
import { FailureContext, Hypothesis, HypothesisStatus, Evidence } from './domain';
import { ILLMProvider } from './llm';

export class Verifier {
  constructor(private llm: ILLMProvider) {}

  async verify(
    hypothesis: Hypothesis,
    evidenceFor: Evidence[],
    evidenceAgainst: Evidence[],
    context: FailureContext
  ): Promise<HypothesisStatus> {
    const prompt = `
You are an independent Verifier.
Context:
${JSON.stringify(context, null, 2)}

Hypothesis: ${hypothesis.statement}

Evidence For:
${JSON.stringify(evidenceFor, null, 2)}

Evidence Against:
${JSON.stringify(evidenceAgainst, null, 2)}

Does the evidence objectively support the hypothesis?
Return JSON only:
{"status": "SUPPORTED" | "REJECTED" | "INCONCLUSIVE"}
`;

    const response = await this.llm.generate({
      systemPrompt: 'You independently verify evidence strictly based on facts.',
      userPrompt: prompt
    });

    try {
      const cleaned = response.trim().replace(/^```json/, '').replace(/```$/, '').trim();
      const parsed = JSON.parse(cleaned);
      if (parsed.status === 'SUPPORTED') return HypothesisStatus.SUPPORTED;
      if (parsed.status === 'REJECTED') return HypothesisStatus.REJECTED;
      return HypothesisStatus.INCONCLUSIVE;
    } catch {
      return HypothesisStatus.INCONCLUSIVE;
    }
  }
}
